### PR TITLE
[WIP] Add reset password form

### PR DESCRIPTION
**Pull request by LordSimal**
## What does this PR improve?

This PR adds a form to request a "Reset my password" email which is linked on the login page.

image

image

After entering a valid registered email you should get an email with a reset password link. After clicking on that you should get:

image

Now if you enter 2 identical "secure" passwords (see `src/Form/ChangePasswordFormType.php`) you should get back to the login form with a confirm message showing you, that the password has been reset.

image

## What did I do?

I basically ran `bin/console make:reset-password` and followed its instructions.

After that I adjusted all generated `templates/reset_password` templates to
- fit/use the bolt core base templates
- transform english text to translatable text

Also I added the ability to set specific configs via the `config/bolt/config.yaml`.

In there you can set the following settings:

```
reset_password_settings:
    mail_from: "do-not-reply@dummy.com"
    mail_name: "Bolt CMS"
    mail_subject: "Your password reset request"
    mail_template: "reset_password/email.html.twig"
```

Also to mention is the fact, that I didn't enable the "This email is not registered function" due to security reasons.
See the `'There was a problem handling your password reset request - %s',` line in the `src/Controller/ResetPasswordController.php`

## What is missing

Since I couldn't get my mail server to allow sending mails via the local environment I couldn't test the mail which should be sent to the user after a successful request.

Also I was not able to translate the text inside this email so I would be very happy if someone can help me there.

### CODE REVIEW

File: src/.preload.php [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?php
```
Comment (bobdenotter) [inline src/.preload.php:1]: We don't need this file ^_^

File: src/Controller/ResetPasswordController.php [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    public function __construct(ResetPasswordHelperInterface $resetPasswordHelper, Config $config)
+    {
+        $this->resetPasswordHelper = $resetPasswordHelper;
+        $this->config = $config;
+    }
+
+    /**
+     * Display & process form to request a password reset.
+     *
+     * @Route("", name="app_forgot_password_request")
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:45]:
```suggestion
     * @Route("", name="bolt_forgot_password_request")
```

File: src/Controller/ResetPasswordController.php [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    /**
+     * Confirmation page after a user has requested a password reset.
+     *
+     * @Route("/check-email", name="app_check_email")
+     */
+    public function checkEmail(): Response
+    {
+        // We prevent users from directly accessing this page
+        if (! $this->canCheckEmail()) {
+            return $this->redirectToRoute('app_forgot_password_request');
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:73]:
```suggestion
            return $this->redirectToRoute('bolt_forgot_password_request');
```

File: src/Controller/ResetPasswordController.php [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        try {
+            $user = $this->resetPasswordHelper->validateTokenAndFetchUser($token);
+        } catch (ResetPasswordExceptionInterface $e) {
+            $this->addFlash('reset_password_error', sprintf(
+                'There was a problem validating your reset request - %s',
+                $e->getReason()
+            ));
+
+            return $this->redirectToRoute('app_forgot_password_request');
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:109]:
```suggestion
            return $this->redirectToRoute('bolt_forgot_password_request');
```

File: src/Controller/ResetPasswordController.php [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+        return $this->render('reset_password/request.html.twig', [
+            'requestForm' => $form->createView(),
+        ]);
+    }
+
+    /**
+     * Confirmation page after a user has requested a password reset.
+     *
+     * @Route("/check-email", name="app_check_email")
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:67]:
```suggestion
     * @Route("/check-email", name="bolt_check_email")
```

File: src/Controller/ResetPasswordController.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+        return $this->render('reset_password/check_email.html.twig', [
+            'tokenLifetime' => $this->resetPasswordHelper->getTokenLifetime(),
+        ]);
+    }
+
+    /**
+     * Validates and process the reset URL that the user clicked in their email.
+     *
+     * @Route("/reset/{token}", name="app_reset_password")
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:84]:
```suggestion
     * @Route("/reset/{token}", name="bolt_reset_password")
```

File: src/Controller/ResetPasswordController.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        $user = $this->getDoctrine()->getRepository(User::class)->findOneBy([
+            'email' => $emailFormData,
+        ]);
+
+        // Marks that you are allowed to see the app_check_email page.
+        $this->setCanCheckEmailInSession();
+
+        // Do not reveal whether a user account was found or not.
+        if (! $user) {
+            return $this->redirectToRoute('app_check_email');
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:154]:
```suggestion
            return $this->redirectToRoute('bolt_check_email');
```

File: src/Controller/ResetPasswordController.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        ]);
+    }
+
+    private function processSendingPasswordResetEmail(string $emailFormData, MailerInterface $mailer): RedirectResponse
+    {
+        $user = $this->getDoctrine()->getRepository(User::class)->findOneBy([
+            'email' => $emailFormData,
+        ]);
+
+        // Marks that you are allowed to see the app_check_email page.
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:149]:
```suggestion
        // Marks that you are allowed to see the bolt_check_email page.
```

File: src/Controller/ResetPasswordController.php [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        // Do not reveal whether a user account was found or not.
+        if (! $user) {
+            return $this->redirectToRoute('app_check_email');
+        }
+
+        try {
+            $resetToken = $this->resetPasswordHelper->generateResetToken($user);
+        } catch (ResetPasswordExceptionInterface $e) {
+            // If you want to tell the user why a reset email was not sent, uncomment
+            // the lines below and change the redirect to 'app_forgot_password_request'.
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:161]:
```suggestion
            // the lines below and change the redirect to 'bolt_forgot_password_request'.
```

File: src/Controller/ResetPasswordController.php [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+            // Caution: This may reveal if a user is registered or not.
+            //
+            // $this->addFlash('reset_password_error', sprintf(
+            //     'There was a problem handling your password reset request - %s',
+            //     $e->getReason()
+            // ));
+
+            return $this->redirectToRoute('app_check_email');
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:169]:
```suggestion
            return $this->redirectToRoute('bolt_check_email');
```

File: src/Controller/ResetPasswordController.php [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            ->subject($config['mail_subject'])
+            ->htmlTemplate($config['mail_template'])
+            ->context([
+                'resetToken' => $resetToken,
+                'tokenLifetime' => $this->resetPasswordHelper->getTokenLifetime(),
+            ]);
+
+        $mailer->send($email);
+
+        return $this->redirectToRoute('app_check_email');
```
Comment (bobdenotter) [inline src/Controller/ResetPasswordController.php:187]:
```suggestion
        return $this->redirectToRoute('bolt_check_email');
```

File: templates/reset_password/check_email.html.twig [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        </div>
+
+        <div class="card">
+            <div class="card-header">
+                <i class="fa fa-lock" aria-hidden="true"></i>
+                {{ 'reset_password.request_header'|trans }}
+            </div>
+            <div class="card-body">
+                <p>{{ 'reset_password.check_email_sent_text_1'|trans({'%hours%': tokenLifetime|date('g')}) }}</p>
+                <p>{{ 'reset_password.check_email_sent_text_2'|trans({'%tryagain%': '<a href="' ~ path('app_forgot_password_request') ~ '">try again</a>' }) | raw }}</p>
```
Comment (bobdenotter) [inline templates/reset_password/check_email.html.twig:36]:
```suggestion
                <p>{{ 'reset_password.check_email_sent_text_2'|trans({'%tryagain%': '<a href="' ~ path('bolt_forgot_password_request') ~ '">try again</a>' }) | raw }}</p>
```

File: templates/reset_password/email.html.twig [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<h1>{{ 'reset_password.email_title'|trans }}</h1>
+{# TBD: Hi! #}
+
+<p>{{ 'reset_password.email_description'|trans }}</p>
+{# TBD: To reset your password, please visit the following link #}
+
+<a href="{{ url('app_reset_password', {token: resetToken.token}) }}">{{ url('app_reset_password', {token: resetToken.token}) }}</a>
```
Comment (bobdenotter) [inline templates/reset_password/email.html.twig:7]:
```suggestion
<a href="{{ url('bolt_reset_password', {token: resetToken.token}) }}">{{ url('bolt_reset_password', {token: resetToken.token}) }}</a>
```

File: templates/reset_password/request.html.twig [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+                <a href="{{ path('bolt_login') }}" class="btn btn-primary">
+                    <i class="fas fa-sign-in-alt"></i>
+                    {{ 'action.log_in'|trans }}
+                </a>
+                {{ form_end(requestForm) }}
+            </div>
+        </div>
+
+    </div>
+{% endblock %}
```
Comment (bobdenotter) [inline templates/reset_password/request.html.twig:54]:
```suggestion
{% endblock %}

```

File: templates/security/login.html.twig [hunk 15]
```diff
@@ -70,6 +74,10 @@
                         <i class="fas fa-sign-in-alt" aria-hidden="true"></i>
                         {{ 'action.log_in'|trans }}
                     </button>
+                    <a href="{{ path('app_forgot_password_request') }}" class="btn btn-primary">
```
Comment (bobdenotter) [inline templates/security/login.html.twig:77]:
```suggestion
                    <a href="{{ path('bolt_forgot_password_request') }}" class="float-right btn btn-tertiary">
```

File: templates/reset_password/reset.html.twig [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+        <div class="card">
+            <div class="card-header">
+                <i class="fa fa-lock" aria-hidden="true"></i>
+                {{ 'reset_password.reset_header'|trans }}
+            </div>
+            <div class="card-body">
+                {{ form_start(resetForm) }}
+                {{ form_row(resetForm.plainPassword) }}
+                <button class="btn btn-primary">{{ 'reset_password.reset_btn'|trans }}</button>
```
Comment (bobdenotter) [inline templates/reset_password/reset.html.twig:37]:
```suggestion
                <button class="float-right btn btn-tertiary">{{ 'reset_password.reset_btn'|trans }}</button>
```

**Review by bobdenotter — CHANGES_REQUESTED** [review]

**Comment by bobdenotter** [comment]
I've fixed some low-hanging fruit. Mainly updating the name of the routes. 

We'll need to ensure that we tell people who're upgrading to run the following:

```
bin/console do:sc:up --force
```

For some reason it's not sending emails on my test-server either. Even though I have SMTP set up. I'm going to dig in some more.

File: templates/reset_password/request.html.twig [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+                {{ 'reset_password.request_header'|trans }}
+            </div>
+            <div class="card-body">
+                {{ form_start(requestForm, {'attr': {'class': 'reset-password__form'}}) }}
+                {{ form_row(requestForm.email) }}
+                <div class="reset-password__description">
+                    {{ 'reset_password.request_description'|trans }}
+                </div>
+                <button class="btn btn-primary">{{ 'reset_password.request_send'|trans }}</button>
+                <a href="{{ path('bolt_login') }}" class="btn btn-primary">
```
Comment (bobdenotter) [inline templates/reset_password/request.html.twig:45]:
```suggestion
                <a href="{{ path('bolt_login') }}" class="float-right btn btn-tertiary">
```

File: src/.preload.php (hunk 1 above)
Comment (LordSimal) [reply src/.preload.php:1]: 👍 has been deleted

File: src/Controller/ResetPasswordController.php (hunk 2 above)
Comment (LordSimal) [reply src/Controller/ResetPasswordController.php:45]: changed that 👍

File: src/Controller/ResetPasswordController.php (hunk 3 above)
Comment (LordSimal) [reply src/Controller/ResetPasswordController.php:73]: changed that 👍

File: src/Controller/ResetPasswordController.php (hunk 4 above)
Comment (LordSimal) [reply src/Controller/ResetPasswordController.php:109]: changed that 👍

File: src/Controller/ResetPasswordController.php (hunk 5 above)
Comment (LordSimal) [reply src/Controller/ResetPasswordController.php:67]: changed that 👍

File: src/Controller/ResetPasswordController.php (hunk 11 above)
Comment (LordSimal) [reply src/Controller/ResetPasswordController.php:187]: changed that 👍

File: templates/reset_password/check_email.html.twig (hunk 12 above)
Comment (LordSimal) [reply templates/reset_password/check_email.html.twig:36]: changed that 👍

File: templates/reset_password/reset.html.twig (hunk 16 above)
Comment (LordSimal) [reply templates/reset_password/reset.html.twig:37]: changed that 👍

File: templates/reset_password/request.html.twig (hunk 17 above)
Comment (LordSimal) [reply templates/reset_password/request.html.twig:45]: changed that 👍

File: templates/reset_password/request.html.twig (hunk 14 above)
Comment (LordSimal) [reply templates/reset_password/request.html.twig]: yes my lord 🙇 😄

**Comment by LordSimal** [comment]
- Fixed all app_ routes and changed them to bolt_ routes.
- Added missing translation for E-Mail template

**Comment by I-Valchev** [comment]
@LordSimal @bobdenotter wohoo, finally a password reset feature 🥳

This may be coming in too late, but afaik so far we've not used Symfony forms in core, insteading building the frontend forms with html/twig, e.g. if i look at `login.html.twig`.

For consistency, would it be better to do that too?

**Comment by LordSimal** [comment]
@I-Valchev so instead of building forms like
```
{{ form_start(requestForm, {'attr': {'class': 'reset-password__form'}}) }}
{{ form_row(requestForm.email) }}
...
{{ form_end(requestForm) }}
```

you would like to built it like 

```
<form action="{{ path('bolt_login') }}" method="post" class="" id="login-form">
    <input type="hidden" name="_csrf_token" value="{{ csrf_token('authenticate') }}"/>
    {% include '@bolt/_partials/fields/text.html.twig' with {
        'label' : 'label.username_or_email'|trans,
        'name' : 'username',
        'value' : last_username,
        'disabled' : false,
        'placeholder': 'placeholder.username_or_email'|trans,
        'autofocus': true,
    } %}
</form>
```

and therefore have more control over the form, or am I wrong?

**Comment by I-Valchev** [comment]
@LordSimal exactly 😅

In that case, do we need `ChangePasswordFormType.php`?

**Comment by bobdenotter** [comment]
@LordSimal @I-Valchev Yes, I mostly agree. For long-term maintainability it'd be good if we use the same way of constructing our forms throughout.

On the other hand, this _is_ the de-facto symfony way, and we should perhaps consider moving other forms that are _not_ "Edit Content" as well. Like, 'login', and 'edit profile'. 

So, yeah, i'm a bit conflicted if we should ask @LordSimal to refactor this PR, or if we should add a TODO item for some of those other ones in core right now.. 🤔

**Comment by LordSimal** [comment]
I haven't looked so deeply into the Bolt Core yet but how many "internal" forms are built without Symfony Forms?

**Comment by bobdenotter** [comment]
From the top of my head: 

 - Login
 - Reset password
 - Edit (own) profile
 - Add / Edit user
 - Edit Content (should remain as-is)
 - Edit Media 
 - Edit Configuration / Template / Text files (basically just a CodeMirror instance)
 
Out of scope, imho:  
 - API explorer (straight from API Platform)
 - Translations (Straight from PHPTranslate)
 - And there's a few loose 'search', 'filter', etc inputs, that aren't "real" forms.

**Comment by bobdenotter** [comment]
After giving this some more thought, with @I-Valchev, i think we should keep this as-is. 

Let's instead see if we can migrate some of the "non SF-forms" forms to SF-forms as well, instead. (as separate PR's / tasks)

**Comment by LordSimal** [comment]
well then I will first try to do that with the login form (in a separate PR as you said)

**Review by bobdenotter — APPROVED** [review]
I'm going to merge this in to the `feature/reset-password` branch. That way we can test it properly, in order to include it in the 4.2 release!

**Comment by LordSimal** [comment]
Just wanted to remind @bobdenotter that this feature still is out there waiting to be released 😉  
Just noticed it when installing 4.2 that its not there 😭

**Pull request merged**